File: bbtest/steps/journal_steps.rb
require_relative 'placeholders'

step "snapshot :account version :count should be" do |account, version, expectation|
  (tenant, account) = account.split('/')
  actual = Journal.account_snapshot(tenant, account, version)
  expectation = JSON.parse(expectation)

  expect(actual["version"]).to eq(expectation["version"])
  expect(actual["balance"]).to eq(expectation["balance"])
  expect(actual["promised"]).to eq(expectation["promised"])
  expect(actual["accountName"]).to eq(expectation["accountName"])
  expect(actual["isBalanceCheck"]).to eq(expectation["isBalanceCheck"])
  expect(actual["currency"]).to eq(expectation["currency"])
  expect(actual["promiseBuffer"]).to match_array(expectation["promiseBuffer"])
end

step "transaction :id of :tenant should be" do |id, tenant, expectation|
  transaction = Journal.transaction_data(tenant, id)
  expectation = JSON.parse(expectation)

  expect(transaction["id"]).to eq(expectation["id"]) unless expectation["id"].nil?

  expectation["transfers"].each { |e|
    found = false
    transaction["transfers"].each { |t|
      same = true
      same &&= t["id"] == e["id"] unless e["id"].nil?
      same &&= t["credit"] == e["credit"] unless e["credit"].nil?
      same &&= t["debit"] == e["debit"] unless e["debit"].nil?
      same &&= t["valueDate"] == e["valueDate"] unless e["valueDate"].nil?
      same &&= t["amount"] == e["amount"] unless e["amount"].nil?
      same &&= t["currency"] == e["currency"] unless e["currency"].nil?

      if same
        found = true
        break
      end
    }
    raise "#{e} not found in #{transaction}" unless found
  }
end

step "transaction :id state of :tenant should be :state" do |id, tenant, state|
  expect(Journal.transaction_state(tenant, id)).to eq(state)
end

File: bbtest/steps/journal_steps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from behave import *
import ssl
import urllib.request
import json
import os
from datetime import datetime


@then('snapshot {tenant}/{account} version {version} should be')
def check_account_snapshot(context, tenant, account, version):
  path = '/data/t_{}/account/{}/snapshot/{}'.format(tenant, account, version.zfill(10))
  assert os.path.isfile(path), 'file {} does not exists'.format(path)

  actual = dict()
  with open(path, 'r') as fd:
    lines = fd.readlines()
    lines = [line.strip() for line in lines]

    actual.update({
      'isBalanceCheck': 'true' if lines[0][-1] != "F" else 'false',
      'format': lines[0][4:-2],
      'currency': lines[0][:3],
      'accountName': account,
      'version': version,
      'balance': lines[1],
      'promised': lines[2],
      'promiseBuffer': ' '.join(lines[3:-2])
    })

  for row in context.table:
    assert row['key'] in actual, 'key {} is missing in {}'.format(row['key'], actual)
    assert actual[row['key']] == row['value'], 'expected {} got {}'.format(row['value'], actual[row['key']])


@then('transaction {transaction} state of {tenant} should be {status}')
def check_transaction_state(context, tenant, transaction, status):
  path = '/data/t_{}/transaction/{}'.format(tenant, transaction)
  assert os.path.isfile(path), 'file {} does not exists'.format(path)

  actual = dict()
  with open(path, 'r') as fd:
    lines = fd.readlines()
    lines = [line.strip() for line in lines]
    assert lines[0] == status, 'expected {} but got {}'.format(status, lines[0])


@then('transaction {transaction} of {tenant} should be')
def check_transaction_state(context, tenant, transaction):
  path = '/data/t_{}/transaction/{}'.format(tenant, transaction)
  assert os.path.isfile(path), 'file {} does not exists'.format(path)

  actual = []
  with open(path, 'r') as fd:
    lines = fd.readlines()
    lines = [line.strip() for line in lines]
    state = lines[0]
    for transfer in lines[1:]:
      item = transfer.split(' ')
      actual.append({
        'credit': {
          'name': item[2],
          'tenant': item[1],
        },
        'debit':{
          'name': item[4],
          'tenant': item[3],
        },
        'amount': item[6],
        'currency': item[7],
        #'valueDate': item[5],
      })
  actual = sorted(actual)

  expected = []
  for row in context.table:
    tmp = dict()
    if row.get('credit'):
      (t, a) = row['credit'].split('/')
      tmp['credit'] = {
        'name': a,
        'tenant': t,
      }
    if row.get('debit'):
      (t, a) = row['debit'].split('/')
      tmp['debit'] = {
        'name': a,
        'tenant': t,
      }
    if row.get('amount'):
      tmp['amount'] = row['amount']
    if row.get('currency'):
      tmp['currency'] = row['currency']
    expected.append(tmp)
  expected = sorted(expected)

  assert expected == actual
